nfc/NfcCx/Cx/driver.cpp

```diff
@@ -1132,3 +1134,109 @@ NfcCxUnbindClient(
 Done:
     TRACE_FUNCTION_EXIT_NTSTATUS(LEVEL_VERBOSE, status);
 }
+
+NTSTATUS
+NfcCxEvtReleaseHardwareControl(
+    _In_ PNFCCX_DRIVER_GLOBALS NfcCxGlobals,
+    _In_ WDFDEVICE Device
+)
+/*++
+
+Routine Description:
+
+This routine is called by the CX client to make CX driver
+release the NFCC access.
+
+Arguments:
+
+Device - WDF device to initialize
+
+Return Value:
+
+NTSTATUS
+
+--*/
+{
+    NTSTATUS status = STATUS_SUCCESS;
+    PNFCCX_FDO_CONTEXT fdoContext;
+
+    TRACE_FUNCTION_ENTRY(LEVEL_VERBOSE);
+
+    if (!VerifyPrivateGlobals(NfcCxGlobals)) {
+        TRACE_LINE(LEVEL_ERROR, "Invalid CX global pointer");
+        status = STATUS_INVALID_PARAMETER;
+        goto Done;
+    }
+
+    fdoContext = NfcCxFdoGetContext(Device);
+
+    if (fdoContext->RFInterface == NULL) {
+        TRACE_LINE(LEVEL_ERROR, "CX not initialized");
+        status = STATUS_INVALID_DEVICE_STATE;
+        goto Done;
+    }
+
+    status = NfcCxFdoDeInitialize(fdoContext);
+
+Done:
+
+    TRACE_FUNCTION_EXIT_NTSTATUS(LEVEL_VERBOSE, status);
+    return status;
+}
+
+NTSTATUS
+NfcCxEvtReacquireHardwareControl(
+    _In_ PNFCCX_DRIVER_GLOBALS NfcCxGlobals,
+    _In_ WDFDEVICE Device
+)
+/*++
+
+Routine Description:
+
+This routine is called by the CX client to make CX driver
+reacquire the NFCC access.
+
+Arguments:
+
+Device - WDF device to initialize
+
+Return Value:
+
+NTSTATUS
+
+--*/
+{
+    NTSTATUS status = STATUS_SUCCESS;
+    PNFCCX_FDO_CONTEXT fdoContext;
+
+    TRACE_FUNCTION_ENTRY(LEVEL_VERBOSE);
+
+    if (!VerifyPrivateGlobals(NfcCxGlobals)) {
+            TRACE_LINE(LEVEL_ERROR, "Invalid CX global pointer");
+            status = STATUS_INVALID_PARAMETER;
+            goto Done;
+    }
+    fdoContext = NfcCxFdoGetContext(Device);
+
+    if (fdoContext->RFInterface == NULL) {
+        TRACE_LINE(LEVEL_ERROR, "CX not initialized");
+        status = STATUS_INVALID_DEVICE_STATE;
+        goto Done;
+    }
+
+    status = NfcCxFdoInitialize(fdoContext);
+    if (STATUS_SUCCESS == status)
+    {
+        status = NfcCxRFInterfaceUpdateDiscoveryState(fdoContext->RFInterface);
+    }
+    else
+    {
+        TRACE_LINE(LEVEL_ERROR, "NfcCxFdoInitialize Failed");
+        status = STATUS_INVALID_DEVICE_STATE;
```

Comment:
Reassigning status to STATUS_INVALID_DEVICE_STATE we are losing information about actual failure in NfcCxFdoInitialize